### ISSUE TITLE
`unused_parens` false positive on cast with match expression

### ISSUE THREAD

**Issue by ehuss** [rust-lang/rust#88519]
I tried this code:

```rust
pub fn foo() -> i32 {
    let x = 1;
    (match x {
        _ => x,
    } as i32)
}
```

This generates a warning about unused parentheses:

```
warning: unnecessary parentheses around block return value
 --> src/lib.rs:3:5
  |
3 | /     (match x {
4 | |         _ => x
5 | |     } as i32)
  | |_____________^
  |
  = note: `#[warn(unused_parens)]` on by default
help: remove these parentheses
  |
3 ~     match x {
4 +         _ => x
5 +     } as i32
  |
```

However, applying the suggestion causes it to fail to compile:

```
error: expected expression, found `as`
 --> src/lib.rs:5:7
  |
5 |     } as i32
  |       ^^ expected expression
  |
help: parentheses are required to parse this as an expression
  |
3 ~     (match x {
4 |         _ => x,
5 ~     }) as i32
  |
```

Found in the 2021 crater run for:
* https://crater-reports.s3.amazonaws.com/pr-87190-3/try%23a7a572ce3edd6d476191fbfe92c9c1986e009b34/reg/embly-wrapper-0.0.2/log.txt

### Meta

`rustc --version --verbose`:
```
rustc 1.56.0-nightly (5eacec9ec 2021-08-28)
binary: rustc
commit-hash: 5eacec9ec7e112a0de1011519a57c45586d58414
commit-date: 2021-08-28
host: x86_64-apple-darwin
release: 1.56.0-nightly
LLVM version: 13.0.0
```

**Comment by ehuss** [comment]
Another example:

```rust
(if val == 1 { 0 } else { val } as u64)
```

https://crater-reports.s3.amazonaws.com/pr-87190-3/try%23a7a572ce3edd6d476191fbfe92c9c1986e009b34/gh/ashtuchkin.rypt/log.txt

**Comment by chenyukang** [comment]
As the hint wrote, the parentheses should bed added in a different position: 

```rust
pub fn foo() -> i32 {
    let x = 1;
    **(match x {
        _ => x,
    }) as i32**
}
```

**Comment by chenyukang** [comment]
The second is: 

```rust
pub fn foo_bar(val: i32) -> i32 {
    (if val == 1 { 0 } else { val }) as i32
}
```
